File: hanimo_rag/llm/openai_compat.py
"""OpenAI-compatible LLM provider."""
from __future__ import annotations

import httpx

from hanimo_rag.config import Config
from hanimo_rag.llm.base import LLMBase


class OpenAICompatLLM(LLMBase):
    """Works with OpenAI, Together, Groq, or any OpenAI-compatible endpoint."""

    def __init__(self, config: Config) -> None:
        self.model = config.model_name
        self.base_url = config.openai_base_url.rstrip("/")
        self.api_key = config.openai_api_key
        self.temperature = config.temperature
        self.max_tokens = config.max_tokens

        if not self.api_key:
            raise ValueError(
                "OpenAI API key is required. Set OPENAI_API_KEY env var "
                "or pass openai_api_key to Config."
            )

    async def generate(self, prompt: str, system: str = "") -> str:
        messages: list[dict[str, str]] = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": self.temperature,
            "max_tokens": self.max_tokens,
        }

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

        async with httpx.AsyncClient(timeout=httpx.Timeout(120.0)) as client:
            resp = await client.post(
                f"{self.base_url}/chat/completions",
                json=payload,
                headers=headers,
            )
            resp.raise_for_status()
            data = resp.json()

        choices = data.get("choices", [])
        if not choices:
            return ""
        return choices[0].get("message", {}).get("content", "")

File: js/src/llm/openai-compat.ts
/**
 * hanimo-rag v2 — OpenAI-compatible LLM Implementation
 */

import { LLMBase } from './base.js'

interface ChatCompletionResponse {
  choices: Array<{
    message: {
      role: string
      content: string
    }
  }>
}

export class OpenAICompatLLM extends LLMBase {
  private baseUrl: string
  private model: string
  private apiKey: string
  private temperature: number
  private maxTokens: number

  constructor(options: {
    baseUrl: string
    model: string
    apiKey: string
    temperature: number
    maxTokens: number
  }) {
    super()
    this.baseUrl = options.baseUrl.replace(/\/$/, '')
    this.model = options.model
    this.apiKey = options.apiKey
    this.temperature = options.temperature
    this.maxTokens = options.maxTokens
  }

  async generate(prompt: string, system?: string): Promise<string> {
    const messages: Array<{ role: string; content: string }> = []

    if (system) {
      messages.push({ role: 'system', content: system })
    }
    messages.push({ role: 'user', content: prompt })

    const url = `${this.baseUrl}/v1/chat/completions`

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }
    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`
    }

    let response: Response
    try {
      response = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify({
          model: this.model,
          messages,
          temperature: this.temperature,
          max_tokens: this.maxTokens,
        }),
      })
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err)
      throw new Error(
        `Failed to connect to OpenAI-compatible API at ${this.baseUrl}. Error: ${message}`,
      )
    }

    if (!response.ok) {
      const body = await response.text().catch(() => 'unknown error')
      throw new Error(
        `OpenAI API error (${response.status}): ${body}`,
      )
    }

    const data = (await response.json()) as ChatCompletionResponse

    const content = data.choices[0]?.message.content
    if (!content) {
      throw new Error('OpenAI API returned empty response')
    }

    return content
  }
}
